### ISSUE TITLE
Consider deprecating reactive module in favor of SCF support

### ISSUE THREAD

**Issue by philkoch** [spring-cloud/spring-cloud-stream#1559]
Below is the original discussion of the issue with StreamEmmiter which evolve into a solution and confirmation to utilize recently added Spring Cloud Function support. Further more, given the fact that SCF is at it's core reactive the current reactive module seems to be redundant. So we should consider deprecating it.

=======
Hi all,

I have a problem with the **@StreamEmitter** annotation in spring-cloud-stream **Greenwich.RC1**.
I'm implementing a Source that reads from CouchDB and should store the current sequenceId when the application is shut down. 
For this I wanted to use a method with the **@PreDestroy** annotation. This works fine as long as I don't use the @StreamEmitter annotation. As soon as I annotate my emit-method with it the graceful shutdown seems to be ignored.

This is the source:

```
@EnableBinding(Source.class)
@EnableAutoConfiguration
public class GreetingEmitter {

	@Autowired
	private Database database;

	@StreamEmitter
	public void emit(@Output(Source.OUTPUT) FluxSender output) {

		Changes changes = database.changes()
		                          .includeDocs(true)
		                          .since("0")
		                          .heartBeat(300000)
		                          .continuousChanges();

		Flux<Object> flux = Flux
				.create(fluxSink -> {
					while(changes.hasNext()) {

						ChangesResult.Row feed = changes.next();

						if(!feed.isDeleted()) {
							fluxSink.next(feed.getDoc().toString());
						}
					}
				})
				.publish()
				.autoConnect();

		output.send(flux.log());

	}
```

When I stop(SIGINT) the application in Intellij, or on the console without the @StreamEmitter annotation the output looks shows that the application is shutdown gracefully:
```
2018-12-17 11:36:59.234  INFO 18522 --- [       Thread-3] o.s.i.endpoint.EventDrivenConsumer       : Removing {logging-channel-adapter:_org.springframework.integration.errorLogger} as a subscriber to the 'errorChannel' channel
2018-12-17 11:36:59.234  INFO 18522 --- [       Thread-3] o.s.i.channel.PublishSubscribeChannel    : Channel 'application.errorChannel' has 0 subscriber(s).
2018-12-17 11:36:59.235  INFO 18522 --- [       Thread-3] o.s.i.endpoint.EventDrivenConsumer       : stopped _org.springframework.integration.errorLogger
2018-12-17 11:36:59.245  INFO 18522 --- [       Thread-3] o.s.s.concurrent.ThreadPoolTaskExecutor  : Shutting down ExecutorService 'applicationTaskExecutor'
2018-12-17 11:36:59.277  INFO 18522 --- [       Thread-3] o.s.s.c.ThreadPoolTaskScheduler          : Shutting down ExecutorService 'taskScheduler'
2018-12-17 11:36:59.279  INFO 18522 --- [       Thread-3] o.s.i.monitor.IntegrationMBeanExporter   : Summary on shutdown: nullChannel
2018-12-17 11:36:59.279  INFO 18522 --- [       Thread-3] o.s.i.monitor.IntegrationMBeanExporter   : Summary on shutdown: errorChannel
2018-12-17 11:36:59.279  INFO 18522 --- [       Thread-3] o.s.i.monitor.IntegrationMBeanExporter   : Summary on shutdown: output
2018-12-17 11:36:59.279  INFO 18522 --- [       Thread-3] o.s.i.monitor.IntegrationMBeanExporter   : Summary on shutdown: _org.springframework.integration.errorLogger.handler
```

When I add the Annotation the graceful shutdown does not take place and the output is just 
`Process finished with exit code 130 (interrupted by signal 2: SIGINT)`

Can you tell me if this behaviour is intentional and how I can do some  cleanup tasks on shutdown when using **@StreamEmitter**?

I'm happy to provide more code if needed.
Versions:
spring-boot-starter-parent: 2.1.1.RELEASE
spring-cloud-stream: Greenwich.RC1
java 8 (also tried 11 but did not make a difference)

Thanks!

**Comment by msshitole** [comment]
@olegz and @philkoch 
Just FYI, whereas @PreDestroy annotation works fine with SCF's Spplier<Flux<T>>. The graceful shutdown takes place without any issue.

**Comment by olegz** [comment]
Thank you so much @msshitole for validating it. As you probably already recognizing that introducing Spring Cloud Function (SCF) in Spring Cloud Stream(SCSt) makes SCSt's reactive module kind of redundant as I can't think of any functionality it provides that can not be handled by simply wiring via SCF.

**Comment by philkoch** [comment]
Hi @msshitole @olegz,

thanks for the input, I now changed my project to use spring-cloud-function and indeed, the application now shuts down gracefully.

Here's the code that I currently use:
```
@EnableBinding(Source.class)
@EnableAutoConfiguration
public class GreetingEmitter {

	@Autowired
	private Database database;

	@Bean
	public void emit() {

		Changes changes = database.changes()
		                          .includeDocs(true)
		                          .since("0")
		                          .heartBeat(300000)
		                          .continuousChanges();

		return () -> Flux
				.create(fluxSource -> {
					while(changes.hasNext()) {

						ChangesResult.Row feed = changes.next();

						if(!feed.isDeleted()) {
							fluxSource.next(feed.getDoc().toString());
						}
					}
				})
				.publish()
				.autoConnect();

	}
```
I've added the following property to *application.properties*
```
spring.cloud.stream.function.definition=emit
```

I'm fine using this solution for now. My problem is solved for now.

Thanks!

**Comment by olegz** [comment]
@philkoch @msshitole That is great to hear and a pretty nice confirmation on using SCF support in place of the old reactive module.
So, instead of closing this issue, i'll change it's subject to something along the lines of "Consider deprecating reactive module". This way this discussion and subsequent confirmation serves as justification.

**Comment by olegz** [comment]
Resolved for 2.2.x
